### ISSUE TITLE
Misc Reduce Scatter Bug Fixes/Improvements

### ISSUE THREAD

**Issue by SeanNijjar** [tenstorrent/tt-metal#11335]
While enabling the multi-buffer per channel feature for EDM and debugging reduce scatter hangs, I stumbled upon several areas of improvement. These are mostly very low priority as they are either not possible to hit currently or they have marginal perf impact compared to other issues.

- [ ] Fix latent bug for computing max worker slice size bytes:
```
// TODO: need to make this take the packet size as arg instead of `edm_channel_buffer_size` because
// `edm_channel_buffer_size` can be shrunk
static std::size_t compute_maximum_worker_slice_in_bytes(
```

- [ ] fix computing max num workers (num units per scatter dim)
```
uint32_t per_step_dim_size = input_tensor.get_legacy_shape()[scatter_split_dim] / ring_size;
uint32_t input_tensor_num_units_per_scatter_dim =
        per_step_dim_size / tt::constants::TILE_WIDTH;  // TODO: find the divisibility based on layout
// Should be ...:
// std::size_t input_tensor_n_elems_per_slice = input_tensor.volume() / ring_size;
// uint32_t input_tensor_num_units_per_scatter_dim =
//     input_tensor_n_elems_per_slice / (tt::constants::TILE_WIDTH * tt::constants::TILE_HEIGHT);
```  

- [ ] Truncate CB size in case of very large Erisc buffers:
```
    std::size_t cb_num_pages_per_packet = cw_per_link_edm_builders.at(0).get_eth_buffer_size_bytes() / op_config.get_page_size();
    std::size_t ideal_cb_num_pages = cb_num_pages_per_packet * 2;
    std::size_t cb_num_pages = ideal_cb_num_pages;
    while (cb_num_pages > 32) {
        cb_num_pages /= 2;
    }
```
Needed only for single or double channel EDM invocation

- [ ] Optimize EDM handshake to avoid context switches so eagerly:
```
FORCE_INLINE void eth_setup_handshake(std::uint32_t handshake_register_address, bool is_sender) {
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[4] = 1;
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[5] = 1;
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[6] = 0x1c0ffee1;
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[7] = 0x1c0ffee2;

    erisc_info->channels[0].receiver_ack = 0;
    for (uint32_t i = 1; i < eth_l1_mem::address_map::MAX_NUM_CONCURRENT_TRANSACTIONS; i++) {
        erisc_info->channels[i].bytes_sent = 0;
        erisc_info->channels[i].receiver_ack = 0;
    }
    *(volatile tt_l1_ptr uint32_t *)handshake_register_address = 0;
    if (is_sender) {
        eth_wait_receiver_done();
        {
        // uint32_t count = 0;
        // uint32_t max = 2000000000;
        // while (eth_bytes_are_available_on_channel(0)) {
        //     count ++;
        //     if (count > max) {
        //         run_routing();
        //     }
        //     asm volatile("nop");
        // }
        // count = 0;
        // while (eth_txq_is_busy()) {
        //     count ++;
        //     if (count > max) {
        //         run_routing();
        //     }
        //     asm volatile("nop");
        // }
        // count = 0;
        }

        eth_send_bytes(handshake_register_address, handshake_register_address, 16);
        {
        // count = 0;
        // while (eth_txq_is_busy()) {
        //     count ++;
        //     if (count > max) {
        //         run_routing();
        //     }
        //     asm volatile("nop");
        // }
        // eth_send_done();
        // count = 0;
        }
        eth_wait_for_receiver_done();
        {
        // while (eth_bytes_are_available_on_channel(0)) {
        //     count ++;
        //     if (count > max) {
        //         run_routing();
        //     }
        //     asm volatile("nop");
        // }
        }
    } else {
        eth_wait_for_bytes(16);
        {
        // uint32_t count = 0;
        // uint32_t max = 2000000000;
        // while (erisc_info->channels[0].bytes_sent != 16) {
        //     count ++;
        //     if (count > max) {
        //         run_routing();
        //     }
        //     asm volatile("nop");
        // }

        // count = 0;
        // while (eth_txq_is_busy()) {
        //     count ++;
        //     if (count > max) {
        //         run_routing();
        //     }
        //     asm volatile("nop");
        // }
        }
        eth_receiver_channel_done(0);
    }
}
```

- [ ] Further, split it into halves so the sender side can initiate the handshake as early as possible (instead of only after args are consumed:
```

FORCE_INLINE void eth_setup_handshake_first_half(std::uint32_t handshake_register_address, bool is_sender) {
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[4] = 1;
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[5] = 1;
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[6] = 0x1c0ffee1;
    reinterpret_cast<volatile tt_l1_ptr uint32_t *>(handshake_register_address)[7] = 0x1c0ffee2;

    erisc_info->channels[0].receiver_ack = 0;
    for (uint32_t i = 1; i < eth_l1_mem::address_map::MAX_NUM_CONCURRENT_TRANSACTIONS; i++) {
        erisc_info->channels[i].bytes_sent = 0;
        erisc_info->channels[i].receiver_ack = 0;
    }
    *(volatile tt_l1_ptr uint32_t *)handshake_register_address = 0;
    if (is_sender) {
        uint32_t count = 0;
        uint32_t max = 2000000000;
        // eth_wait_receiver_done();
        while (eth_bytes_are_available_on_channel(0)) {
            count ++;
            if (count > max) {
                run_routing();
            }
            asm volatile("nop");
        }
        count = 0;
        while (eth_txq_is_busy()) {
            count ++;
            if (count > max) {
                run_routing();
            }
            asm volatile("nop");
        }
        count = 0;
        eth_send_bytes(handshake_register_address, handshake_register_address, 16);
    } else {

    }
}


FORCE_INLINE void eth_setup_handshake_second_half(std::uint32_t handshake_register_address, bool is_sender) {

    if (is_sender) {
        uint32_t count = 0;
        uint32_t max = 2000000000;
        count = 0;
        while (eth_txq_is_busy()) {
            count ++;
            if (count > max) {
                run_routing();
            }
            asm volatile("nop");
        }
        eth_send_done();
        count = 0;
        // eth_wait_for_receiver_done();
        while (eth_bytes_are_available_on_channel(0)) {
            count ++;
            if (count > max) {
                run_routing();
            }
            asm volatile("nop");
        }
    } else {
        uint32_t count = 0;
        uint32_t max = 2000000000;
        // eth_wait_for_bytes(16);
        while (erisc_info->channels[0].bytes_sent != 16) {
            count ++;
            if (count > max) {
                run_routing();
            }
            asm volatile("nop");
        }

        count = 0;
        while (eth_txq_is_busy()) {
            count ++;
            if (count > max) {
                run_routing();
            }
            asm volatile("nop");
        }
        eth_receiver_channel_done(0);

    }

}
```
